aws/retry/middleware.go

```diff
@@ -0,0 +1,177 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	"github.com/awslabs/smithy-go/transport/http"
+)
+
+type retryMetadata struct {
+	AttemptNum       int
+	AttemptTime      time.Time
+	MaxAttempts      int
+	AttemptClockSkew time.Duration
+}
+
+type retryMetadataKey struct{}
+
+type RetryMiddleware struct {
+	Retryer Retryer
+}
+
+func (r RetryMiddleware) ID() string {
+	return "Retry Middleware"
+}
+
+func (r RetryMiddleware) Name() string {
+	return r.ID()
+}
+
+func (r RetryMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	const invocationIDHeader = "amz-sdk-invocation-id"
+
+	req := in.Request.(*http.Request)
+
+	// TODO: Don't see a compelling reason to make this a separate middleware that makes this available via context
+	invocationID, err := sdk.UUIDVersion4()
+	if err != nil {
+		return out, err
+	}
+
+	req.Header.Set(invocationIDHeader, invocationID)
+
+	var attemptNum, retryCount int
+	var attemptClockSkew time.Duration
+
+	maxAttempts := r.Retryer.MaxAttempts()
+
+	relRetryToken := r.Retryer.GetInitialToken()
+	for {
+		attemptNum++
+
+		ctx = context.WithValue(ctx, retryMetadataKey{}, retryMetadata{
+			AttemptNum:       attemptNum,
+			AttemptTime:      sdk.NowTime(),
+			MaxAttempts:      maxAttempts,
+			AttemptClockSkew: attemptClockSkew,
+		})
+
+		out, reqErr := next.HandleFinalize(ctx, in)
+
+		relRetryToken(reqErr)
+		if reqErr == nil {
+			return out, nil
+		}
+
+		retryable := r.Retryer.IsErrorRetryable(reqErr)
+		if !retryable {
+			return out, err
+		}
+
+		if maxAttempts > 0 && attemptNum >= maxAttempts {
+			err = &aws.MaxAttemptsError{
+				Attempt: attemptNum,
+				Err:     err,
+			}
+			return out, err
```

Comment:
don't need to solve it in this PR, but this is a good example of a place where we should think about how to include response metadata upstream.

Especially since this is returning both `out` response and `err` generally this is an anti-pattern that we should avoid. It probably would be better to have a way of decorating the error with response and retry metadata. Potentially a wrapper retry always wraps any error returned with, (e.g. via a defer)